src/Eval.hs

```diff
@@ -52,38 +59,34 @@ import Debug.Trace
 -- it gets real weird with laziness. (Note to the note: this code is mostly a
 -- remnant of us using StateT, and might not be necessary anymore since we
 -- switched to more explicit state-passing.)
-eval :: Context -> XObj -> IO (Context, Either EvalError XObj)
-eval ctx xobj@(XObj o i t) =
+eval :: Context -> XObj -> Bool -> IO (Context, Either EvalError XObj)
```

Comment:
I'd prefer an custom sumtype over a Bool here to avoid "boolean blindness".

Reply:
Good point! Added a simple sumtype to capture this instead.